src/stream/src/executor/backfill.rs

```diff
@@ -0,0 +1,380 @@
+                                        let mut buffer = vec![];
+                                        mem::swap(&mut upstream_chunk_buffer, &mut buffer);
+                                        for chunk in buffer {
+                                            if self.current_pos.is_some() {
+                                                yield Message::Chunk(Self::mapping_chunk(
+                                                    Self::mark_chunk(
+                                                        chunk,
+                                                        self.current_pos.as_ref().unwrap(),
+                                                        &table_pk_indices,
+                                                    ),
+                                                    &upstream_indices,
+                                                ));
+                                            }
+                                        }
+                                    }
+
+                                    // Update snapshot read epoch.
+                                    snapshot_read_epoch = barrier.epoch.prev;
+
+                                    yield Message::Barrier(barrier);
+
+                                    if checkpoint {
+                                        self.progress
+                                            .update(snapshot_read_epoch, snapshot_read_epoch);
+                                        // Break the for loop and start a new snapshot read stream.
+                                        break;
+                                    }
+                                }
+                                Message::Chunk(chunk) => {
+                                    // Buffer the upstream chunk.
+                                    upstream_chunk_buffer.push(chunk.compact());
+                                }
+                                Message::Watermark(_) => todo!(
+                                    "https://github.com/risingwavelabs/risingwave/issues/6042"
+                                ),
+                            }
+                        }
+                        Either::Right(msg) => {
+                            match msg? {
+                                None => {
+                                    // End of the snapshot read stream.
+                                    // We need to set current_pos to the maximum value or do not
+                                    // mark the chunk anymore, otherwise, we will ignore some rows
+                                    // in the buffer. Here we choose to never mark the chunk.
+                                    // Consume with the renaming stream buffer chunk without mark.
+                                    let mut buffer = vec![];
+                                    mem::swap(&mut upstream_chunk_buffer, &mut buffer);
+                                    for chunk in buffer {
```

Comment:
```suggestion
                                    for chunk in upstream_chunk_buffer.drain(..) {
```
Same for other occurrences.